### PR TITLE
Chore: DB Drizzle Schema

### PR DESCRIPTION
**Pull request by bassiounix**
initial schema.

some functionalities are missing on the drizzle side .. future work arounds are needed

### CODE REVIEW

**Review by G0maa — CHANGES_REQUESTED** [review]
3ash ya5oia

File: .gitignore [hunk 1]
```diff
@@ -140,3 +141,6 @@ dist
 .yarn/build-state.yml
 .yarn/install-state.gz
 .pnp.*
+
+# db
+migrations/
```
[the remark this answers is not in the document]
Comment (bassiounix) [reply .gitignore:146]: this was a mistake, on it

File: apps/api/drizzle.config.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import "dotenv/config";
+import { defineConfig } from "drizzle-kit";
+
+export default defineConfig({
+    dialect: "postgresql",
+    schema: "./src/drizzle/schema",
+    out: "./src/drizzle/migrations",
+    dbCredentials: {
+        url: String(process.env.DATABASE_URL),
```
[the remark this answers is not in the document]
Comment (bassiounix) [reply apps/api/drizzle.config.ts:9]: An example file of .env is ok for the whole suite of configs

File: apps/api/src/drizzle/drizzle.module.ts [hunk 3]
```diff
@@ -2,20 +2,22 @@ import { DrizzlePGModule } from "@knaadh/nestjs-drizzle-pg";
...
 
 @Module({
     imports: [
-        DrizzlePGModule.register({
-            tag: "DB_DEV",
-            pg: {
-                connection: "client",
-                config: {
-                    connectionString: ConnectionString,
+        DrizzlePGModule.registerAsync({
```
[the remark this answers is not in the document]
Comment (bassiounix) [reply apps/api/src/drizzle/drizzle.module.ts:8]: We could, however this one made the heavy lifting for us, as in, we would do the same thing as the package did.
It gives a wrapper (adapter) around Drizzle to use it in the Nest style of doing things (& Nest IoC).

File: apps/api/src/drizzle/schema/server.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import { integer, pgTable, serial, varchar } from "drizzle-orm/pg-core";
+import { inviteLinks } from "./invite-links";
+import { serverUsers } from "./server-user";
+
+export const servers = pgTable("servers", {
+    id: serial("id").primaryKey(),
+    name: varchar("name").notNull(),
+    image_url: varchar("image_url").notNull(),
+
+    owner_id: integer("owner_id").notNull(), // TODO: figure out the right relation for the owner (the egg and the chicken dilemma)
```
[the remark this answers is not in the document]
Comment (bassiounix) [reply apps/api/src/drizzle/schema/server.ts:11]: The problem is wether the owner_id is a server user or a regular user (refer back to ERD)
as in, 2 tables are dependent on each other, so this gives an error in TypeScript on my side

File: apps/api/package.json [hunk 5]
```diff
@@ -7,45 +7,38 @@
     "license": "UNLICENSED",
     "scripts": {
         "build": "nest build",
-        "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
         "start": "nest start",
-        "dev": "nest start --watch",
+        "dev": "nest start --exec \"node --env-file=../../.env.backend\" --watch",
```
Comment (alikehel) [inline apps/api/package.json:11]: Correct this.

File: apps/api/package.json [hunk 6]
```diff
@@ -7,45 +7,38 @@
...
         "@nestjs/config": "^3.2.3",
-        "@nestjs/config": "^3.2.3",
         "@nestjs/core": "^10.0.0",
         "@nestjs/graphql": "^12.2.0",
         "@nestjs/platform-express": "^10.0.0",
         "argon2": "^0.41.1",
-        "nestjs-zod": "^3.0.0",
         "drizzle-orm": "^0.33.0",
         "graphql": "^16.9.0",
+        "nestjs-zod": "^3.0.0",
```
Comment (alikehel) [inline apps/api/package.json:36]: Deprecated package.

File: apps/api/src/drizzle/drizzle.module.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (alikehel) [reply apps/api/src/drizzle/drizzle.module.ts:8]: > Are we really absolutely sure about this module?
> 
> Can't we code it ourselves instead? @abdo0nasser @alikehel @bassiounix

I'm not a fan of using it.

**Comment by alikehel** [comment]
Dont forget to pull from main and then
1. resolve any merge conflict other than the lockfile
2. run `pnpm i`
3. run `git add --update`
4. run `git commit`
5. close the opened file
6. push ur changes

File: apps/api/src/drizzle/schema/server.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (G0maa) [reply apps/api/src/drizzle/schema/server.ts]: @bassiounix Just user, when he creates the server he is then added as a member too.

File: apps/api/src/drizzle/drizzle.module.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (G0maa) [reply apps/api/src/drizzle/drizzle.module.ts:8]: @bassiounix 
I still do not get what's the difference between it, and creating a custom global injectable drizzle service :thinking:

File: apps/api/drizzle.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (G0maa) [reply apps/api/drizzle.config.ts:9]: @bassiounix yes, it is. But later we will have a database for testing that runs on e.g. a different port. Which will load a different `.env` file.

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (G0maa) [reply .gitignore]: @bassiounix resolve if fixed

File: apps/api/drizzle.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bassiounix) [reply apps/api/drizzle.config.ts:9]: Then the only change will be the env variable value, that is, found in the new env files

Also dotenv is in use for this particular case, and settings can be applied easily

File: apps/api/src/drizzle/drizzle.module.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bassiounix) [reply apps/api/src/drizzle/drizzle.module.ts:8]: Nothing different other than the work is already done in the package, as in every other package out there!

File: apps/api/src/drizzle/schema/server.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bassiounix) [reply apps/api/src/drizzle/schema/server.ts]: Yes but not prior to that (order of operations)

Fixed anyway

File: apps/api/src/drizzle/drizzle.module.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (G0maa) [reply apps/api/src/drizzle/drizzle.module.ts:8]: @bassiounix Did you try/research a different approach? This [one](https://wanago.io/2024/05/20/api-nestjs-drizzle-orm-postgresql/) in particular. My problem with this package is it's not an official integration, and isn't widely used.

File: apps/api/package.json (hunk 5 above)
Comment (bassiounix) [reply apps/api/package.json]: done

File: apps/api/package.json (hunk 6 above)
Comment (bassiounix) [reply apps/api/package.json]: done

File: apps/api/src/drizzle/drizzle.module.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bassiounix) [reply apps/api/src/drizzle/drizzle.module.ts:8]: still the same things but written in a package .. that's all really

**Pull request merged**